File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/javascript/segment-tree.js
/**
 * JavaScript Implementation: Segment Tree
 * Invariant: Range Query O(log N), Point/Range Update O(log N)
 */
export class SegmentTree {
  constructor() {
    this.items = [];
  }

  // TODO: Implement verified invariant
  insert(val) {
    this.items.push(val);
  }

  get size() {
    return this.items.length;
  }
}

File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/python/segment-tree.py
"""
Python Implementation: Segment Tree
Invariant: Range Query O(log N), Point/Range Update O(log N)
"""
from typing import Generic, TypeVar, List, Optional

T = TypeVar('T')

class SegmentTree(Generic[T]):
    def __init__(self) -> None:
        self._items: List[T] = []

    # TODO: Implement core operational method
    def insert(self, value: T) -> None:
        self._items.append(value)

    def __len__(self) -> int:
        return len(self._items)
